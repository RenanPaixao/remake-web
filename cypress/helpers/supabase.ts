import { createClient } from '@supabase/supabase-js'

const e2eSupabase = createClient(import.meta.env.VITE_SUPABASE_URL, import.meta.env.VITE_SUPABASE_SERVICE_ROLE_KEY, {
  auth: {
    autoRefreshToken: false,
    persistSession: false
  }
})

export { e2eSupabase }
