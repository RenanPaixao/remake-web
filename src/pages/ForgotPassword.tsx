import { useFormik } from 'formik'
import {
  Button,
  ButtonGroup,
  Center,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Heading,
  Input,
  Modal,
  ModalBody,
  ModalContent,
  ModalFooter,
  ModalHeader,
  ModalOverlay,
  Text
} from '@chakra-ui/react'
import { useNavigate } from 'react-router-dom'
import * as Yup from 'yup'
import { hasFormikError } from '../utils/hasFormikError.ts'
import { supabase } from '../utils/supabase.ts'
import { useErrorToast } from '../hooks/toast/useErrorToast.tsx'
import { useState } from 'react'
import { LanguageButton } from '../components/LanguageButton/LanguageButton.tsx'

const validationSchema = Yup.object({
  email: Yup.string().email('Invalid email address').required('Required').max(50, 'Email must be at most 50 characters')
})

type FormValues = Yup.InferType<typeof validationSchema>

export const ForgotPassword = () => {
  const navigate = useNavigate()
  const toastError = useErrorToast()
  const [isOpen, setIsOpen] = useState(false)

  const formik = useFormik<FormValues>({
    initialValues:{
      email: ''
    },
    onSubmit: async values => {
      const { error } = await supabase.auth.resetPasswordForEmail(values.email)

      if(error) {
        toastError({ description: error.message })
        console.error(error)
        return
      }

      setIsOpen(true)
    },
    validationSchema,
    validateOnBlur: true
  })

  /**
   * Go to the login page.
   */
  function goToLogin() {
    navigate('/login')
  }

  /**
   * Close the modal.
   */
  function closeModal() {
    setIsOpen(false)
  }

  return <>
    <Center pt={28} flexDirection={'column'}>
      <Heading size={'sm'} mb={8}>
      Enter your email address that we will send you a recovery email.
      </Heading>
      <form onSubmit={formik.handleSubmit}>
        <FormControl
          mt={4}
          isRequired
          isInvalid={hasFormikError<FormValues>('email', formik.touched, formik.errors)}
        >
          <FormLabel htmlFor="email">Email</FormLabel>
          <Input
            aria-label={'Email'}
            name={'email'}
            type={'email'}
            onChange={formik.handleChange}
            onBlur={formik.handleBlur}
            value={formik.values.email}
          />
          <FormErrorMessage>{formik.errors.email}</FormErrorMessage>
        </FormControl>

        <ButtonGroup mt={8} spacing={10} isDisabled={formik.isSubmitting}>
          <Button variant={'outline'} isLoading={formik.isSubmitting} onClick={goToLogin}>Login</Button>
          <Button
            colorScheme={'blue'}
            type={'submit'}
            isLoading={formik.isSubmitting}
            isDisabled={Object.keys(formik.errors).length > 0}
          >
          Send Recovery Email
          </Button>
        </ButtonGroup>
      </form>
      <LanguageButton mt={'200px'} variant={'ghost'}/>
    </Center>

    <Modal
      isOpen={isOpen}
      onClose={closeModal}
      closeOnOverlayClick={false}
      closeOnEsc={false}
    >
      <ModalOverlay />
      <ModalContent>
        <ModalHeader>Recovery Email Sent</ModalHeader>
        <ModalBody>
          <Text>We have sent you an email to reset your password. Check your inbox!</Text>
        </ModalBody>
        <ModalFooter>
          <Button colorScheme={'blue'} onClick={goToLogin}>Login</Button>
        </ModalFooter>
      </ModalContent>
    </Modal>
  </>
}
