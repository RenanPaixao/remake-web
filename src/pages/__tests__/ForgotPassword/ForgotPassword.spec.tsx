import { fireEvent } from '@testing-library/dom'
import { act } from '@testing-library/react'
import { expect } from 'vitest'
import { customRender } from '../../../../tests/test-utils.tsx'
import { ForgotPassword } from '../../ForgotPassword.tsx'
import { supabase } from '../../../utils/supabase.ts'

const mockResetPasswordForEmail = vi.fn()
const toastMock = vi.fn()

supabase.auth.resetPasswordForEmail = mockResetPasswordForEmail

describe('ForgotPassword', () => {
  beforeEach(() => {
    vi.mock('react-router-dom', async importOriginal => {
      const actual: object = await importOriginal()
      return {
        ...actual,
        useNavigate: () => vi.fn()
      }
    })

    vi.mock('@chakra-ui/react', async importOriginal => {
      const actual: object = await importOriginal()
      return {
        ...actual,
        useToast: () => toastMock
      }
    })
  })

  it('should render', async () => {
    const { container } = customRender(<ForgotPassword/>)

    expect(container).toMatchSnapshot()
  })

  it('should send email and show confirmation modal', async () => {
    const { getByRole } = customRender(<ForgotPassword/>)

    const email = 'email@email.com'

    const emailInput = getByRole('textbox', { name: /Email/i })
    const submitButton = getByRole('button', { name: /Send Recovery Email/i })

    await act(async ()=> {
      fireEvent.change(emailInput, { target: { value: email  } })
      fireEvent.click(submitButton)
    })

    expect(mockResetPasswordForEmail).toHaveBeenCalledWith(email)
    expect(toastMock).not.toBeCalled()
  })

  it('should shown an error case something wrong happens when sending the recovery email', async () => {
    const { getByRole } = customRender(<ForgotPassword/>)

    const email = 'email@email.com'
    const error = new Error('Something went wrong')

    mockResetPasswordForEmail.mockReturnValueOnce({ error })

    const emailInput = getByRole('textbox', { name: /Email/i })
    const submitButton = getByRole('button', { name: /Send Recovery Email/i })

    await act(async ()=> {
      fireEvent.change(emailInput, { target: { value: email  } })
      fireEvent.click(submitButton)
    })

    expect(toastMock).toBeCalledTimes(1)
    expect(toastMock).toHaveBeenCalledWith(expect.objectContaining({
      description: error.message,
      status: 'error'
    }))
  })
})
