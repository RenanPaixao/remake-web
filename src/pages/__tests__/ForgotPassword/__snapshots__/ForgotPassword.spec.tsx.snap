// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ForgotPassword > should render 1`] = `
<div>
  <div
    class="css-1n29es9"
  >
    <h2
      class="chakra-heading css-1mgeuu9"
    >
      Enter your email address that we will send you a recovery email.
    </h2>
    <form>
      <div
        class="chakra-form-control css-1mn9ptq"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="email"
          id="field-:r0:-label"
        >
          Email
          <span
            aria-hidden="true"
            class="chakra-form__required-indicator css-1tfjd1n"
            role="presentation"
          >
            *
          </span>
        </label>
        <input
          aria-label="Email"
          aria-required="true"
          class="chakra-input css-1cjy4zv"
          id="field-:r0:"
          name="email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="chakra-button__group css-1p2xg5l"
        data-orientation="horizontal"
        role="group"
      >
        <button
          class="chakra-button css-15jenze"
          type="button"
        >
          Login
        </button>
        <button
          class="chakra-button css-1tduvxh"
          type="submit"
        >
          Send Recovery Email
        </button>
      </div>
    </form>
    <button
      aria-controls="menu-list-:r2:"
      aria-expanded="false"
      aria-haspopup="menu"
      class="chakra-button chakra-menu__menu-button css-bxcf73"
      id="menu-button-:r2:"
      type="button"
    >
      <span
        class="css-xl71ch"
      >
        ğŸ‡ºğŸ‡¸ English
      </span>
      <span
        class="chakra-button__icon css-1hzyiq5"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
          />
        </svg>
      </span>
    </button>
    <div
      class="css-r6z5ec"
      style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
    >
      <div
        aria-orientation="vertical"
        class="chakra-menu__menu-list css-1kfu8nn"
        id="menu-list-:r2:"
        role="menu"
        style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
        tabindex="-1"
      >
        <button
          class="chakra-menu__menuitem css-18esm8n"
          data-index="0"
          id="menu-list-:r2:-menuitem-:r3:"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
           
          ğŸ‡ºğŸ‡¸ English
        </button>
        <button
          class="chakra-menu__menuitem css-18esm8n"
          data-index="1"
          id="menu-list-:r2:-menuitem-:r4:"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
          ğŸ‡§ğŸ‡· PortuguÃªs
        </button>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
